spec/index.md

```diff
@@ -1298,78 +1298,80 @@ In the reply callback of a [inter-canister method call](#system-api-call), the a
 ### Overview of imports {#system-api-imports}
 
 The following sections describe various System API functions, also referred to as system calls, which we summarize here.
-
-    ic0.msg_arg_data_size : () -> i32;                                          // I U Q CQ Ry CRy F
-    ic0.msg_arg_data_copy : (dst : i32, offset : i32, size : i32) -> ();        // I U Q CQ Ry CRy F
-    ic0.msg_caller_size : () -> i32;                                            // *
-    ic0.msg_caller_copy : (dst : i32, offset: i32, size : i32) -> ();           // *
-    ic0.msg_reject_code : () -> i32;                                            // Ry Rt CRy CRt
-    ic0.msg_reject_msg_size : () -> i32;                                        // Rt CRt
-    ic0.msg_reject_msg_copy : (dst : i32, offset : i32, size : i32) -> ();      // Rt CRt
-
-    ic0.msg_reply_data_append : (src : i32, size : i32) -> ();                  // U Q CQ Ry Rt CRy CRt
-    ic0.msg_reply : () -> ();                                                   // U Q CQ Ry Rt CRy CRt
-    ic0.msg_reject : (src : i32, size : i32) -> ();                             // U Q CQ Ry Rt CRy CRt
-
-    ic0.msg_cycles_available : () -> i64;                                       // U Rt Ry
-    ic0.msg_cycles_available128 : (dst : i32) -> ();                            // U Rt Ry
-    ic0.msg_cycles_refunded : () -> i64;                                        // Rt Ry
-    ic0.msg_cycles_refunded128 : (dst : i32) -> ();                             // Rt Ry
-    ic0.msg_cycles_accept : (max_amount : i64) -> (amount : i64);               // U Rt Ry
-    ic0.msg_cycles_accept128 : (max_amount_high : i64, max_amount_low: i64, dst : i32)
-                           -> ();                                               // U Rt Ry
-
-    ic0.cycles_burn128 : (amount_high : i64, amount_low : i64, dst : i32) -> ();               // I G U Ry Rt C T
-
-    ic0.canister_self_size : () -> i32;                                         // *
-    ic0.canister_self_copy : (dst : i32, offset : i32, size : i32) -> ();       // *
-    ic0.canister_cycle_balance : () -> i64;                                     // *
-    ic0.canister_cycle_balance128 : (dst : i32) -> ();                          // *
-    ic0.canister_status : () -> i32;                                            // *
-    ic0.canister_version : () -> i64;                                           // *
-
-    ic0.msg_method_name_size : () -> i32;                                       // F
-    ic0.msg_method_name_copy : (dst : i32, offset : i32, size : i32) -> ();     // F
-    ic0.accept_message : () -> ();                                              // F
-
-    ic0.call_new :                                                              // U CQ Ry Rt CRy CRt T
-      ( callee_src  : i32,
+The notation `I ∈ {i32, i64}` means that there are actually two corresponding System API functions for the two integer types `I`: `i32` and `i64`.
```

Comment:
Maybe the overloading could lead to difficulties in Rust and other languages that do not support overloading. This can probably be solved by specifying a Wasm binding attribute but I am not sure how easy this is.

Reply:
I guess you shouldn't need to use the two versions at the same time so conditional compilation should do the job.

Reply:
I agree, importing the API functions in a canister is probably easy as the compiler can selectively import only one version of the overloaded functions as needed for the corresponding memory address space. The only thing I am not sure is related to the IC implementation as it would have to export the overloaded functions, because a canister may decide to mix both 32-bit and 64-bit IC calls. Probably, wasm bindings can solve this mapping in Rust.

Reply:
Actually, the replica implementation wouldn't allow overloading system API functions (adding an overloaded function [here](https://github.com/dfinity/ic/blob/d51f0f58f899e404be9bc0b60a0d35df53cb2553/rs/embedders/src/wasmtime_embedder/system_api.rs#L434) crashes the replica).

Reply:
Thanks for the analysis, Martin. 

Maybe, the other idea would work where the replica would only expose one version of the function with the pointer type dependent on the bit-width of the canister's main memory?

Reply:
> with the pointer type dependent on the bit-width of the canister's main memory?

How about letting the canister choose the bit-width as an optional argument (with 32-bit default) to install_code?

Reply:
A nice design property of Wasm Memory64 is that it is automatically backwards-compatible to Wasm32. No canister option is needed on the IC to enable it and it can be dynamically detected in the runtime system. Ideally, we could select the IC system API bit-width also automatically, without having an explicit install option. I believe this would be simpler for the users and the implementation and it would also be a little bit more consistent (the API bit-width would always match the memory bit-width). If I am not mistaken, introducing a install option for the system API bit-width would be more complicated as it needs to be stored as additional configuration in the canister state (to be available for subsequent IC calls on that canister). This would probably require more effort than the `wasm_memory_persistence` upgrade option that is only temporarily needed during the upgrade and does not need to be stored beyond the install message.

Reply:
I don't really see why a canister would be using a mix of the two types. So we could probably infer which types should be used from the type of the memory defined in the Wasm module?

Reply:
> I don't really see why a canister would be using a mix of the two types. So we could probably infer which types should be used from the type of the memory defined in the Wasm module?

That's get's my vote.

Reply:
+1 on this as well.

Reply:
Sounds good. I've updated this PR to derive the value of `I` from the bit-width of the memory.